src/components/tooltip/GCMarker.js

```diff
@@ -423,80 +430,103 @@ function _treeInsert(
-    if (node.value) {
-      // We delete the value to change this node from a leaf to a branch.
-      delete node.value;
-    }
+    node.leaf = false;
     _treeInsert(node.branches, path, phase);
   } else {
     // Make the new node leaf node.
-    if (node.value) {
+    if (node.value.name !== '(temp)') {
       console.error(
-        'Duplicate phases in _treeInsert in MarkerTooltipContents.js'
+        `Duplicate phase ${phase.name} in _treeInsert in MarkerTooltipContents.js`
       );
       return;
     }
     node.value = phase;
+    node.leaf = true;
   }
 }
 
-function _treeGetLeaves(
-  tree: Map<string, PhaseTreeNode>
-): Array<PhaseTimeTuple> {
-  const leaves = [];
+function _treeMapNodes(
+  tree: Map<string, PhaseTreeNode>,
+  f: (node: PhaseTreeNode) => any,
+  results: Array<any> = []
+): Array<any> {
   for (const node of tree.values()) {
-    if (node.value) {
-      leaves.push(node.value);
-    } else {
-      leaves.push(..._treeGetLeaves(node.branches));
+    const result = f(node);
+    if (result !== undefined) {
+      results.push(result);
+    }
+    if (!node.leaf) {
+      _treeMapNodes(node.branches, f, results);
     }
   }
-  return leaves;
+  return results;
 }
 
 function _filterInterestingPhaseTimes(
   rawPhases: PhaseTimes<Microseconds>,
   numSelect: number
 ): Array<PhaseTimeTuple> {
-  let phaseTimes = _makePhaseTimesArray(rawPhases);
+  const phaseTimes = _makePhaseTimesArray(rawPhases);
 
   /*
-   * Select only the leaf phases.
+   * Build the tree.
    */
   const tree = new Map();
   for (const phase of phaseTimes) {
     const components = phase.name.split('.');
     _treeInsert(tree, components, phase);
   }
-  phaseTimes = _treeGetLeaves(tree);
 
   /*
-   * Of those N leaf phases, select the M most interesting phases by
-   * determining the threshold we want to stop including phases at.
-   *
-   * Calculate the threshold by sorting the list of times in asscending
-   * order, then slicing off all the low items and looking at the first
-   * item.
+   * For every non-leaf phase, add in a (self-time) leaf with the time not
+   * accounted for by its children.
    */
-  const sortedPhaseTimes = phaseTimes
-    .map(pt => pt.time)
-    // Descending order
-    .sort((a, b) => b - a);
-  const threshold = sortedPhaseTimes[numSelect];
+  _treeMapNodes(
+    tree,
+    ({ leaf, branches, value: { time: totalTime, name } }) => {
+      if (!leaf) {
+        let remaining = totalTime;
+        for (const { value } of branches.values()) {
+          remaining -= value.time;
+        }
+        const selfTimeName = `${name} (self-time)`;
+        branches.set(name, {
+          branches: new Map(),
+          leaf: true,
+          value: { name: selfTimeName, time: remaining },
+        });
+      }
+    }
+  );
+
+  // Select the N most interesting leaf phases, in order of descending time.
+  const sortedPhaseTimes = _treeMapNodes(tree, node =>
+    node.leaf ? node.value : undefined
```

Comment:
No. But why do you think so? I'm asking because I thought exactly the same thing, and went to the trouble of implementing a complete path tracking mechanism -- and then discovered that these values are storing the complete paths already. So I'm wondering why I thought that.

Anyway, I have a live profiler window here that has full paths in the tooltip, so I'm confident that it's good.

Reply:
I'm sorry, i wrote that comment before I'd read the rest of the patch and saw how it worked.  I meant to come back and delete the draft comment but forgot.  Please disregard.